QEfficient/generation/llm_generator.py

```diff
@@ -0,0 +1,298 @@
+# -----------------------------------------------------------------------------
+#
+# Copyright (c)  2024 Qualcomm Innovation Center, Inc. All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+#
+# -----------------------------------------------------------------------------
+
+import json
+import os
+import torch
+import numpy as np
+import torch.nn as nn
+
+from typing import Dict, List, Optional, Union
+from threading import Thread
+
+from QEfficient.generation.cloud_infer import QAICInferenceSession
+
+from transformers import (
+    AutoTokenizer,
+    TextIteratorStreamer,
+    TextStreamer,
+    AutoTokenizer,
+    LogitsProcessorList,
+    MinLengthLogitsProcessor,
+    TopKLogitsWarper,
+    TemperatureLogitsWarper,
+    StoppingCriteriaList,
+    MaxLengthCriteria,
+)
+
+
+
+class LLMGenerator:
+    def __init__(
+        self,
+        qpc_path,
+        model_name,
+        device_id: Optional[List[int]] = [0],
+        prompt_len: Optional[int] = 32,
+        ctx_len: Optional[int] = 128,
+        streamer: Optional[Union[TextStreamer, TextIteratorStreamer]] = None,
+        logits_processor: Optional = None,
+        logits_warper: Optional = None,
+    ):
+        self.session = None
+        self.tokenizer = None
+        self.is_first_prompt = False
+        self.model_name = ""
+        self.qpc_path = ""
+        self.device_id = [0]
+        self.curr_cache_index = 0
+        self.ctx_len = ctx_len
+        self.retained_state = True
+        self.write_io_dir = False
+        self.prompt_len = prompt_len
+        self.generated_ids = []
+        self.inputs = None
+        self.stop_indicator = True
+
+        self.qpc_path = (
+            qpc_path if os.path.exists(qpc_path) else OSError(f"{qpc_path} not found !")
+        )
+        self.device_id = device_id
+
+        self.model_name = model_name
+
+        try:
+            self.session = QAICInferenceSession(
+                self.qpc_path, self.device_id, enable_debug_logs=False
+            )
+            if self.retained_state:
+                self.session.skip_buffers(
+                    [x for x in self.session.input_names if x.startswith("past_")]
+                )
+                self.session.skip_buffers(
+                    [
+                        x
+                        for x in self.session.output_names
+                        if x.endswith("_RetainedState")
+                    ]
+                )
+
+        except Exception as err:
+            raise RuntimeError(f"Unable to load qpc on device , {err}")
+
+        try:
+            hf_token = None
+            if os.getenv("HF_TOKEN") is not None:
+                hf_token = os.getenv('HF_TOKEN')
+            tokenizer = AutoTokenizer.from_pretrained(
```

Comment:
Have you tested this on models present in tests? I have seen it failing with config file not found on huggingface, and the work-around is to use `hf_download`, check code in `QEfficient/cloud/infer.py::80-87` you can allow only tokenizer file pattern, as you don't need other model files for tokenizer

Reply:
yes, hf_download are used un-necessarily, even at place where we just need tokenizer still we are downloading all files @ochougul please raise an issue

Reply:
Here, we download all the files while only needing tokenizer, assuming that user must already have all the files in `cache_dir`, in which case, the files won't be downloaded again, as this was how this API is supposed to be used.

If you want to change it to only download tokenizer files when only tokenizer is needed, you are welcome to update it. This is definitely a better way to do this.

You can raise issue internally yourself, no need to ask anybody. Thanks.

Reply:
we cannot assume user will have these files downloaded, IMO just bad design, so didn't move out line 80-87 to main, it will be required in top level api. Consider scaling this solution as aws service and user just wants use qeff apis to run inference on some edge device or device with low space, how will your solution pan out? Happy to discuss more.